addons/l10n_uy/models/res_partner.py

```diff
@@ -0,0 +1,100 @@
+        if partner.name and partner.name != 'False':
+            msg = _(
+                'The %(vat_label)s number [%(wrong_vat)s] for %(partner_label)s does not seem to be valid.'
+                '\nNote: the expected format is %(expected_format)s',
+                vat_label=vat_label,
+                wrong_vat=partner.vat,
+                partner_label=_("partner [%s]", partner.name),
+                expected_format=expected_format,
+            )
+        else:
+            msg = _(
+                'The %(vat_label)s number [%(wrong_vat)s] does not seem to be valid.'
+                '\nNote: the expected format is %(expected_format)s',
+                vat_label=vat_label,
+                wrong_vat=partner.vat,
+                expected_format=expected_format,
+            )
+        return msg
+
+    def _l10n_uy_ci_nie_is_valid(self):
+        """ Check if the partner's CI or NIE number is a valid one.
+
+        CI:
+            1) The ID number is taken up to the second to last position, that is, the first 6 or 7 digits.
+            2) Each digit is multiplied by a different factor starting from right to left, the factors are:
+                2, 9, 8, 7, 6, 3, 4.
+            3) The products obtained are added:
+            4) The base module 10 is calculated on this result to obtain the check digit, expressed in another way,
+            the next number ending in zero is taken that follows the result of the addition (for the example
+            would be 60) subtracting the sum itself: 60 - 59 = 1. The verification digit of the example ID is 1.
+
+            NOTE: If the ID has fewer digits, it is preceded with zeros and the mechanism described above is applied
+
+        NIE:
+            The calculation for the NIE is the same as that used for the CI. The only difference is that we skip the
+            first number
+
+        Both algorithms where extracted from Uruware's Technical Manual (section 9.2 and 9.3)
+
+        Return: False is not valid, True is valid
+        """
+        self.ensure_one()
+
+        # The VAT must consist only numbers (format could have these characters ":., " we can skip them later)
+        invalid_chars = re.findall(r"[^0-9:., \-]", self.vat)
+        if invalid_chars:
+            return False
+
+        ci_nie_number = re.sub('[^0-9]', '', self.vat)
+
+        # we get the validation digit, if NIE doc type we skip the first digit
+        is_nie = self.l10n_latam_identification_type_id.l10n_uy_dgi_code == '1'
+        verif_digit = int(ci_nie_number[-1])
+        ci_nie_number = ci_nie_number[1:-1] if is_nie else ci_nie_number[0:-1]
+
+        # If number is < 7 digits we add 0 to the left
+        ci_nie_number = "%07d" % int(ci_nie_number)
+
+        # If NIE > 7 digits is not valid
+        if len(ci_nie_number) > 7:
+            return False
+
+        verification_vector = (2, 9, 8, 7, 6, 3, 4)
+        num_sum = sum(int(ci_nie_number[i]) * verification_vector[i] for i in range(6))
```

Comment:
Maybe you meant to have `range(7)` here? Otherwise the last `verification_vector`'s item (`4`) is not going to get looped in the num_sum

Reply:
Hi @yosuanicolaus 
In this case I check it again the formula, and it is ok. It will the into account the last element of the array verification_vector is an array of length 7. Using  range(6) we will have numbers between 0 and 6 to access exactly the array positions (from 0 to 6).

Best Regards

Reply:
In python if we do `range(x)` it will generate number from 0 up to `x - 1`, but not the `x`.
For example, if you do this in python in terminal:
```py
[i for i in range(6)]
```
it will return
```py
[0, 1, 2, 3, 4, 5]
```
Which is `6` numbers, not `7`.

... So you should do `range(7)` in this case :)

Reply:
Hi @yosuanicolaus 
Thank you for you comments.
indeed, sorry about that, for some reasons I assumed that range(6) result was [0, 1, 2, 3, 4, 5, 6]
I have updated the PR with the last changes

Best Regards